### PR TITLE
executor: modify lookupTableTask to return merged rows, and improve AppendRows

### LINKED ISSUE #20843: hptc ycsb-worloade: Modify lookupTableTask to return merged rows, and improve AppendRows
## Description

Every row from resultTask has its own chunk, even if their column definitions are same. This makes it impossible to `appendColumn(chk.columns[0], rows.columns[0])`, i.e. append one column of lots of rows
 to another column of the chunk(or the result) at once.

https://github.com/pingcap/tidb/tree/master/executor/distsql.go#L67-L70

https://github.com/pingcap/tidb/blob/master/executor/distsql.go#L1139-L1141

Modify lookupTableTask to return merged rows(maybe a chunk), gc may be improved by less live objects, `AppendRows` may make use of this feature to append data more efficiently. 

The task is to improve AppendRows by trying append one column of multiples rows, in one shot, or several slice `append`.

If, sadly, that the definition of `columns` needs to be modified to get the issue done, the difficulty will be much larger, and should have higher rewards.

## Score

- 600

## Mentor

- @xhebox 

## Sig Slack Channel

You can join #sig-exec on slack in your spare time to discuss and get help with mentors or others.

## Recommended Skills

+ Golang

## Learning Materials

N/A

### PR DESCRIPTION
**Pull request by TszKitLo40**
hptc
### What problem does this PR solve?

Issue Number: close #20843 

Problem Summary:

### What is changed and how it works?

What's Changed:
modify lookupTableTask to return merged rows, and improve AppendRows
How it Works:
Modify lookupTableTask to return merged rows(chunk), gc may be improved by less live objects, AppendRows may make use of this feature to append data more efficiently.

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### Benchmark
500000 records in a single machine
The test command is `go-ycsb run mysql -P ./workloads/workloade -p recordcount=100000 -p operationcount=10000 -p maxscanlength=10000 -p limit=1000 -p mysql.host=127.0.0.1 -p mysql.port=4000 -p mysql.user=root -p mysql.db=test`

```
baseline
INSERT - Takes(s): 9.0, Count: 17, OPS: 1.9, Avg(us): 2269, Min(us): 1101, Max(us): 5876, 99th(us): 6000, 99.9th(us): 6000, 99.99th(us): 6000
SCAN   - Takes(s): 9.9, Count: 229, OPS: 23.1, Avg(us): 43327, Min(us): 2050, Max(us): 99874, 99th(us): 92000, 99.9th(us): 100000, 99.99th(us): 100000

INSERT - Takes(s): 19.0, Count: 28, OPS: 1.5, Avg(us): 2198, Min(us): 1101, Max(us): 5876, 99th(us): 6000, 99.9th(us): 6000, 99.99th(us): 6000
SCAN   - Takes(s): 19.9, Count: 448, OPS: 22.5, Avg(us): 44432, Min(us): 2050, Max(us): 99874, 99th(us): 91000, 99.9th(us): 100000, 99.99th(us): 100000
INSERT - Takes(s): 29.0, Count: 45, OPS: 1.5, Avg(us): 2315, Min(us): 652, Max(us): 5876, 99th(us): 6000, 99.9th(us): 6000, 99.99th(us): 6000
SCAN   - Takes(s): 29.9, Count: 685, OPS: 22.9, Avg(us): 43604, Min(us): 2050, Max(us): 99874, 99th(us): 91000, 99.9th(us): 100000, 99.99th(us): 100000

INSERT - Takes(s): 39.0, Count: 61, OPS: 1.6, Avg(us): 2190, Min(us): 652, Max(us): 5876, 99th(us): 6000, 99.9th(us): 6000, 99.99th(us): 6000
SCAN   - Takes(s): 39.9, Count: 922, OPS: 23.1, Avg(us): 43179, Min(us): 2050, Max(us): 99874, 99th(us): 89000, 99.9th(us): 100000, 99.99th(us): 100000

This branch20843
INSERT - Takes(s): 9.0, Count: 15, OPS: 1.7, Avg(us): 1348, Min(us): 1094, Max(us): 2749, 99th(us): 3000, 99.9th(us): 3000, 99.99th(us): 3000
SCAN  - Takes(s): 9.9, Count: 273, OPS: 27.5, Avg(us): 36506, Min(us): 2662, Max(us): 83279, 99th(us): 81000, 99.9th(us): 84000, 99.99th(us): 84000

INSERT - Takes(s): 19.0, Count: 26, OPS: 1.4, Avg(us): 1272, Min(us): 778, Max(us): 2749, 99th(us): 3000, 99.9th(us): 3000, 99.99th(us): 3000
SCAN  - Takes(s): 19.9, Count: 513, OPS: 25.7, Avg(us): 38888, Min(us): 2394, Max(us): 98823, 99th(us): 83000, 99.9th(us): 99000, 99.99th(us): 99000

INSERT - Takes(s): 29.0, Count: 34, OPS: 1.2, Avg(us): 1282, Min(us): 746, Max(us): 2749, 99th(us): 3000, 99.9th(us): 3000, 99.99th(us): 3000
SCAN  - Takes(s): 29.9, Count: 779, OPS: 26.0, Avg(us): 38399, Min(us): 1749, Max(us): 98823, 99th(us): 83000, 99.9th(us): 99000, 99.99th(us): 99000

INSERT - Takes(s): 39.0, Count: 48, OPS: 1.2, Avg(us): 1304, Min(us): 746, Max(us): 2749, 99th(us): 3000, 99.9th(us): 3000, 99.99th(us): 3000
SCAN  - Takes(s): 39.9, Count: 1031, OPS: 25.8, Avg(us): 38690, Min(us): 1749, Max(us): 98823, 99th(us): 84000, 99.9th(us): 96000, 99.99th(us): 99000
```

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -1161,26 +1192,36 @@ func (w *tableWorker) executeTask(ctx context.Context, task *lookupTableTask) er
 		sort.Sort(task)
 	}
 
-	if handleCnt != len(task.rows) && !util.HasCancelled(ctx) {
+	if handleCnt != task.numRows && !util.HasCancelled(ctx) {
```
Comment (xhebox) [inline executor/distsql.go:1195]: You can add methods GetNumRows and GetRow for task, to share code here.

File: util/chunk/chunk.go [hunk 2]
```diff
@@ -340,6 +340,14 @@ func (c *Chunk) NumRows() int {
 	return c.columns[0].length
 }
 
+// GetNumRows encapsulate NumRows and avoid chunk is nil.
```
Comment (xhebox) [inline util/chunk/chunk.go:343]: Just check nil pointer by the caller rather than here.

**Review by xhebox — CHANGES_REQUESTED** [review]

File: executor/distsql.go [hunk 3]
```diff
@@ -1123,7 +1136,10 @@ func (w *tableWorker) executeTask(ctx context.Context, task *lookupTableTask) er
 	task.memUsage = memUsage
 	task.memTracker.Consume(memUsage)
 	handleCnt := len(task.handles)
-	task.rows = make([]chunk.Row, 0, handleCnt)
+	if w.keepOrder {
```
Comment (wshwsh12) [inline executor/distsql.go:1139]: Why we need to distinguish keep-order or non-keep-order?

File: executor/distsql.go [hunk 4]
```diff
@@ -111,6 +113,17 @@ func (task *lookupTableTask) Swap(i, j int) {
 	task.rows[i], task.rows[j] = task.rows[j], task.rows[i]
 }
 
+func (task *lookupTableTask) buildResultChunk() {
+	if task.numRows <= 0 {
+		return
+	}
+	if task.rows[0].Chunk() != nil {
+		task.chk = chunk.Renew(task.rows[0].Chunk(), task.rows[0].Chunk().NumRows())
```
Comment (b41sh) [inline executor/distsql.go:121]: renew a `Chunk` here will take up extra memory space, which may lead to worse performance.

File: executor/distsql.go [hunk 5]
```diff
@@ -1123,7 +1136,10 @@ func (w *tableWorker) executeTask(ctx context.Context, task *lookupTableTask) er
 	task.memUsage = memUsage
 	task.memTracker.Consume(memUsage)
 	handleCnt := len(task.handles)
-	task.rows = make([]chunk.Row, 0, handleCnt)
+	if w.keepOrder {
+		task.rows = make([]chunk.Row, 0, handleCnt)
+	}
```
Comment (b41sh) [inline executor/distsql.go:1141]: We can build `task.chk` here, if `w.keepOrder` is  false.

**Comment by wshwsh12** [comment]
Could we close this PR now?

**Comment by b41sh** [comment]
> Could we close this PR now?

Optimization performance is not good, we can close this pr.

**Pull request closed without merging**